3-UsingAPIs/Week1/homework/ex3-rollDice.js

```diff
@@ -10,46 +10,47 @@
 ------------------------------------------------------------------------------*/
 
 // TODO Remove callback and return a promise
-function rollDice(callback) {
-  // Compute a random number of rolls (3-10) that the dice MUST complete
-  const randomRollsToDo = Math.floor(Math.random() * 8) + 3;
-  console.log(`Dice scheduled for ${randomRollsToDo} rolls...`);
-
-  const rollOnce = (roll) => {
-    // Compute a random dice value for the current roll
-    const value = Math.floor(Math.random() * 6) + 1;
-    console.log(`Dice value is now: ${value}`);
-
-    // Use callback to notify that the dice rolled off the table after 6 rolls
-    if (roll > 6) {
-      // TODO replace "error" callback
-      callback(new Error('Oops... Dice rolled off the table.'));
-    }
-
-    // Use callback to communicate the final dice value once finished rolling
-    if (roll === randomRollsToDo) {
-      // TODO replace "success" callback
-      callback(null, value);
-    }
-
-    // Schedule the next roll todo until no more rolls to do
-    if (roll < randomRollsToDo) {
-      setTimeout(() => rollOnce(roll + 1), 500);
-    }
-  };
-
-  // Start the initial roll
-  rollOnce(1);
+function rollDice() {
+  return new Promise((resolve, reject) => {
+    // Compute a random number of rolls (3-10) that the dice MUST complete
+    const randomRollsToDo = Math.floor(Math.random() * 8) + 3;
+    console.log(`Dice scheduled for ${randomRollsToDo} rolls...`);
+
+    const rollOnce = (roll) => {
+      // Compute a random dice value for the current roll (39.30 new example)
+      const value = Math.floor(Math.random() * 6) + 1;
+      console.log(`Dice value is now: ${value}`);
+
+      // Use callback to notify that the dice rolled off the table after 6 rolls
+      if (roll > 6) {
+        // TODO replace "error" callback
+        reject(new Error('Oops... Dice rolled off the table.'));
+      }
+
+      // Use callback to communicate the final dice value once finished rolling
+      if (roll === randomRollsToDo) {
+        // TODO replace "success" callback
+        resolve(value);
+      }
+
+      // Schedule the next roll todo until no more rolls to do
+      if (roll < randomRollsToDo) {
+        setTimeout(() => rollOnce(roll + 1), 500);
+      }
+    };
+    rollOnce(1);
+  });
 }
+/*Since "success" and "error" are linked to the loop (roll number) when using callbacks, 
+we keep getting errors and success while the loop continues. In callback, we receive error for every value but in promise, success and error are 
+associated with the resolve and reject. in other words  (they should have same elements i mean if error is an object reject should be also an object.)
+despite the loop continues.  
+*/
```

Comment:
The explanation I was hoping for is something like: once a promise is settled (i.e. resolved or rejected) its outcome can no longer be changed. Subsequent calls to `resolve` and `reject` have no effect then.